### PR TITLE
Add the notion of @_implementationOnly overrides

### PR DESCRIPTION
**Pull request by jrose-apple**
When an `@_implementationOnly` import includes Objective-C categories for existing types, it's useful to be able to override the members provided in those categories without exposing them to clients of the framework being built. Allow this as long as the overriding declaration is marked as `@_implementationOnly` itself, with an additional check that the type of the declaration does not change. (Normally overrides are allowed to change in covariant ways.)

This is definitely not sound when not using `-enable-library-evolution`, and when it is using `-enable-library-evolution` it's probably relying on the fact that overrides of ObjC things are always `dynamic` and that overridable Swift members can never be added in extensions. But the feature is still underscored and there's a definite need for it.

Part of rdar://50827914

### CODE REVIEW

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Comment by jrose-apple** [comment]
Follow-up work: when using `@_implementationOnly`, skip checking the type of a declaration too.

File: lib/Sema/TypeCheckAttr.cpp [hunk 1]
```diff
@@ -2639,6 +2640,41 @@ void AttributeChecker::visitFunctionBuilderAttr(FunctionBuilderAttr *attr) {
...
+
+  auto selfInterfaceTy = VD->getDeclContext()->getDeclaredInterfaceType();
+
+  auto overrideInterfaceTy =
+      selfInterfaceTy->adjustSuperclassMemberDeclType(overridden, VD,
+                                                      baseInterfaceTy);
+  if (!derivedInterfaceTy->isEqual(overrideInterfaceTy)) {
+    TC.diagnose(VD, diag::implementation_only_override_changed_type,
+                overrideInterfaceTy);
+    TC.diagnose(overridden, diag::overridden_here);
```
Comment (slavapestov) [inline lib/Sema/TypeCheckAttr.cpp]: We tend to prefer calling diagnose() on the decl instead, I’m service of eventually removing unnecessary dependencies on the TypeChecker instance

**Review by slavapestov — APPROVED** [review]

File: lib/Sema/TypeCheckAttr.cpp (hunk 1 above)
Comment (jrose-apple) [reply lib/Sema/TypeCheckAttr.cpp]: Oh, I forgot that Decls have a direct `diagnose` now. Will update.

**Comment by jrose-apple** [comment]
@swift-ci Please test Linux

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test macOS

**Comment by jrose-apple** [comment]
@swift-ci Please test Linux

**Comment by jrose-apple** [comment]
Holding off on merging until I've played with this on one of the Apple frameworks using the feature, to make sure it's behaving as expected.

**Comment by jrose-apple** [comment]
@swift-ci Please test

File: lib/Sema/TypeCheckAttr.cpp [hunk 2]
```diff
@@ -2630,6 +2631,69 @@ void AttributeChecker::visitFunctionBuilderAttr(FunctionBuilderAttr *attr) {
...
+    auto derivedInterfaceFuncTy = derivedInterfaceTy->castTo<AnyFunctionType>();
+    derivedInterfaceTy =
+        FunctionType::get(derivedInterfaceFuncTy->getParams(),
+                          derivedInterfaceFuncTy->getResult());
+    auto overrideInterfaceFuncTy =
+        overrideInterfaceTy->castTo<AnyFunctionType>();
+    overrideInterfaceTy =
+        FunctionType::get(overrideInterfaceFuncTy->getParams(),
+                          overrideInterfaceFuncTy->getResult());
+  }
```
Comment (jrose-apple) [inline lib/Sema/TypeCheckAttr.cpp]: I am *not happy* with this whole block. Before, this section of the function was four lines that were a bit verbose but basically said what they did. Now there's a bunch of extra type munging for no obvious reason.

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Pull request merged**